### ISSUE TITLE
Deadlock in statsd.Client.Close()

### ISSUE THREAD

**Issue by danp60** [DataDog/datadog-go#123]
I've encountered what I believe to be a deadlock in the `Close()` logic. When the main goroutine is closing, the mutex for the `Client` is acquired, we send a signal to all other goroutines to stop, then we wait for those goroutines to stop: https://github.com/DataDog/datadog-go/blob/master/statsd/statsd.go#L547-L555
``` golang
func (c *Client) Close() error {
	if c == nil {
		return ErrNoClient
	}
	c.Lock() // <- acquire the lock here
	defer c.Unlock()
	select {
	case <-c.stop:
		return nil
	default:
	}
	close(c.stop)
	c.wg.Wait()
```

However, in the `watch()` goroutine, its possible for the ticker to fire _after_ the main thread has acquired the mutex, but before the `stop` channel is closed. This means that the `watch()` goroutine will also try and acquire the same mutex here: https://github.com/DataDog/datadog-go/blob/master/statsd/statsd.go#L334 and block.
``` golang
func (c *Client) watch() {
	ticker := time.NewTicker(c.flushTime)

	for {
		select {
		case <-ticker.C:
			c.Lock() // <- This blocks because the main goroutine holds the lock
```

Since the main goroutine waits for the `watch()` goroutine to exit here: https://github.com/DataDog/datadog-go/blob/master/statsd/statsd.go#L555, we have a deadlock.

``` golang
func (c *Client) Close() error {
	if c == nil {
		return ErrNoClient
	}
	c.Lock() // <- acquire the lock here
	defer c.Unlock()
	select {
	case <-c.stop:
		return nil
	default:
	}
	close(c.stop)
	c.wg.Wait() # <- This deadlocks because `watch()` is waiting for c.Lock() to be released
```

**Comment by hush-hush** [comment]
This has been fixed in `3.4.1`

**Comment by jared2501** [comment]
Thank you!
